File: python/arxiv_common/src/arxiv_common/ratelimit.py
"""Cross-process arXiv request budget: one request every N seconds across everything we run.

`SharedMinInterval` reserves send slots in Redis through schemas/ratelimit.lua, the same script the
Rust poller and fetcher use, so a Python producer (backfill) can run alongside them. `LocalMinInterval`
is the in-process fallback and the `--local-ratelimit` option.
"""

import os
import time
from functools import lru_cache
from pathlib import Path

import redis
import structlog

DEFAULT_KEY = "arxiv:ratelimit"
# Slots are reserved min + guard apart (same constant as rust/common/src/ratelimit.rs): a sleeper can
# wake a few ms late while the next process lands exactly on its slot, so without a guard consecutive
# sends measure min - jitter. With it the configured interval is a true floor; cost 1.7% at 3 s.
SLOT_GUARD_MS = 50
log = structlog.get_logger()


def script_path() -> Path:
    override = os.environ.get("ARXIV_RATELIMIT_SCRIPT_PATH")
    if override:
        return Path(override)
    # <repo>/python/arxiv_common/src/arxiv_common/ratelimit.py -> <repo>/schemas/ratelimit.lua
    return Path(__file__).resolve().parents[4] / "schemas" / "ratelimit.lua"


@lru_cache(maxsize=1)
def script_source() -> str:
    return script_path().read_text()


class LocalMinInterval:
    """Minimum spacing between requests inside this process only."""

    def __init__(self, min_interval_s: float) -> None:
        self.min = min_interval_s
        self._last: float | None = None

    def wait(self) -> float:
        """Block until this process may send; returns the seconds slept."""
        waited = 0.0
        if self._last is not None:
            waited = max(0.0, self._last + self.min - time.monotonic())
            if waited > 0:
                time.sleep(waited)
        self.mark_sent()
        return waited

    def mark_sent(self) -> None:
        self._last = time.monotonic()


class SharedMinInterval:
    """Reserves the next free slot on a Redis key shared by every arXiv-facing process.

    Reservation is atomic and FIFO (see the script header), so callers never race. If Redis is
    unreachable the call warns and falls back to in-process spacing: a Redis blip must never stop
    a harvest, but must never skip the spacing either.
    """

    def __init__(self, client: redis.Redis, min_interval_s: float, key: str = DEFAULT_KEY) -> None:
        self.client = client
        self.key = key
        self.interval_ms = int(round(min_interval_s * 1000)) + SLOT_GUARD_MS
        self._script = client.register_script(script_source())
        self._fallback = LocalMinInterval(min_interval_s)

    def reserve(self) -> float:
        """Claim the next slot without sleeping; returns the seconds to wait before sending."""
        ms = self._script(keys=[self.key], args=[self.interval_ms])
        return int(ms) / 1000

    def wait(self) -> float:
        """Block until the reserved slot; returns the seconds slept."""
        try:
            waited = self.reserve()
        except redis.exceptions.RedisError as e:
            log.warning("shared rate limiter unavailable; in-process spacing for this request", error=str(e))
            return self._fallback.wait()
        if waited > 0:
            time.sleep(waited)
        self._fallback.mark_sent()
        return waited

File: schemas/ratelimit.lua
-- arXiv request budget: one atomic send-slot reservation per call.
--
-- Contract (implemented by rust/common/src/ratelimit.rs and python/arxiv_common/ratelimit.py):
--   KEYS[1]  budget key, default "arxiv:ratelimit". Holds the most recently reserved send slot in
--            Redis-server milliseconds (redis.call('TIME')), so every process shares one clock and
--            no host clock ever matters.
--   ARGV[1]  minimum spacing between requests, in ms (3000 for arXiv).
--   ARGV[2]  optional key TTL in ms (default 3600000 = 1 h). The TTL is refreshed on every
--            reservation, so a key left behind by a crash expires instead of living forever, while
--            a queue of waiting callers (slot far ahead of now) is never cut short.
--   returns  integer ms the caller must sleep before sending (0 = send now).
--
-- This reserves rather than checks: each call claims slot = max(last + interval, now), stores it, and
-- returns slot - now. Concurrent callers in any process therefore get distinct slots in arrival order
-- (FIFO) and never race or retry. A caller that reserves and then does not send just leaves a gap.
local t = redis.call('TIME')
local now = tonumber(t[1]) * 1000 + math.floor(tonumber(t[2]) / 1000)
local interval = tonumber(ARGV[1])
local ttl = tonumber(ARGV[2]) or 3600000
local last = tonumber(redis.call('GET', KEYS[1]))
local slot = now
if last and last + interval > slot then
  slot = last + interval
end
redis.call('SET', KEYS[1], slot, 'PX', ttl)
return slot - now
